### PR TITLE
Created config.py file

### PR DESCRIPTION
**Pull request by frutkick**
created Config class with configparser as method

### CODE REVIEW

File: ecomap/db/ecomap/CREATE_DB.sql [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
-SOURCE schema/problem_type.sql;
-SOURCE schema/problem.sql;
-SOURCE schema/resource.sql;
-SOURCE schema/role_permission.sql;
-SOURCE schema/role.sql;
-SOURCE schema/user_role.sql;
-SOURCE schema/user.sql;
-SOURCE schema/vote.sql;
-SOURCE schema/page.sql;
+DROP DATABASE IF EXISTS ecomap_db;
```
Comment (lhalam) [inline ecomap/db/ecomap/CREATE_DB.sql]: wrong marge

File: ecomap/db/ecomap/INSERT_DATA.sql [hunk 2]
```diff
@@ -1 +1,7 @@
-SOURCE data/user.sql;
+
```
Comment (lhalam) [inline ecomap/db/ecomap/INSERT_DATA.sql]: wrong marge

File: ecomap/src/python/ecomap/config.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+from ConfigParser import SafeConfigParser
+from time import time
+import logging
+
+password = ['password', 'pass', 'pswd']          # keys which must be string
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Why do you need this line? Could you please explain? What do you think happen if user will set password as argument 'db_pass' in ecomap.conf?

File: ecomap/src/python/ecomap/config.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+logger = logging.getLogger('example')
+logging.basicConfig(level=logging.DEBUG)
+timeToUpdate = 15 * 60                           # 15 minutes
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Please use for method, variable next format:
Constants:
REFRESH_TIME
class name:
Config or ConfigParser
method:
_parse_confs
variable:
self.create_time

File: ecomap/src/python/ecomap/config.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from ConfigParser import SafeConfigParser
```
Comment (lhalam) [inline ecomap/src/python/ecomap/config.py]: code convention

File: ecomap/src/python/ecomap/config.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            cls._instance = object.__new__(cls)  # create it
+        return cls._instance                     # else return it
+
+    def __init__(self, path, logger):
+        self.config = {}                         # dictionary, contains configs
+        self.timeUpdate = None                   # time to update dictionary
+        self.timeCreated = time()                # create time of instance
+        self.path = path                         # path to file (temporary)
+        self.logger = logger
+        self._parseConfs()                       # get configs at start
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Wrong:). You don't need this here

File: ecomap/src/python/ecomap/config.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    def get(self):                               # method which checks if we
+        if self.timeUpdate < time():             # need to update configs
+            self.config = {}                     # nullify configs dictionary
+            self.timeUpdate = time() + timeToUpdate  # set time to update
+            self._parseConfs()                   # parse config file
+        return self.config
+
+    def _parseConfs(self):
+        self.logger.debug('Parsed ecomap.conf at {}'.format(time()))
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Please use for logging next format:
self.logger.debug('Parsed ecomap.conf at %s', time.time())

Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: You need maybe use  info level here.

File: ecomap/src/python/ecomap/config.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        config = SafeConfigParser()             # create config object
+        config.readfp(open(self.path))          # read file
+        sections = config.sections()            # get sections
+        for section in sections:                # for each section
+            for (key, value) in config.items(section):   # for each key/value
+                if value and key.lower() not in password:
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Why do you need this check (not in password)?

File: ecomap/src/python/ecomap/config.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+from ConfigParser import SafeConfigParser
+from time import time
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: please write import time only

File: ecomap/src/python/ecomap/config.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __new__(cls, *args):
+        if not hasattr(cls, '_instance'):        # if there is not instance
+            cls._instance = object.__new__(cls)  # create it
+        return cls._instance                     # else return it
+
+    def __init__(self, path, logger):
+        self.config = {}                         # dictionary, contains configs
+        self.timeUpdate = None                   # time to update dictionary
+        self.timeCreated = time()                # create time of instance
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Why do you need self.timeUpdate and self.timeCreated?

File: ecomap/etc/ecomap.conf [hunk 11]
```diff
@@ -0,0 +1,15 @@
+[ecomap]
+user = admin
+password = 4321
+pool_size = 50
+float = 5.0
```
Comment (ulyktey) [inline ecomap/etc/ecomap.conf]: ??????

File: ecomap/src/python/ecomap/config.py (hunk 3 above)
Comment (padalko) [inline ecomap/src/python/ecomap/config.py]: we can change it using regexp

File: ecomap/src/python/ecomap/config.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+class Config(object):
+
+    def __new__(cls, *args):
+        if not hasattr(cls, '_instance'):        # if there is not instance
+            cls._instance = object.__new__(cls)  # create it
+        return cls._instance                     # else return it
+
+    def __init__(self, path, logger):
+        self.config = {}                         # dictionary, contains configs
+        self.update_time = time.time() + REFRESH_TIME  # time of living
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: self.update_time = 0 :)

File: ecomap/src/python/ecomap/config.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
+from ConfigParser import SafeConfigParser
+import logging
+import time
+import unittest
+
+logger = logging.getLogger('example')
+logging.basicConfig(level=logging.DEBUG)
+REFRESH_TIME = 15 * 60                           # 15 minutes
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: # Comment
REFRESH_TIME = 900

File: ecomap/src/python/ecomap/config.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class Config(object):
+
+    def __new__(cls, *args):
+        if not hasattr(cls, '_instance'):        # if there is not instance
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: http://stackoverflow.com/questions/6760685/creating-a-singleton-in-python,  please read Answers 14: Use a Metaclass

File: ecomap/src/python/ecomap/config.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+import time
+import unittest
+
+logger = logging.getLogger('example')
+logging.basicConfig(level=logging.DEBUG)
+REFRESH_TIME = 15 * 60                           # 15 minutes
+
+
+class Config(object):
+
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Could you please add docstrings to class, methods:
Example:
"""Replaces template placeholder with values.

:param timestamp: formatted date to display
:param priority: priority number
:param priority_name: priority name
:param message: message to display
:returns: formatted string
"""

File: ecomap/src/python/ecomap/config.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
+from ConfigParser import SafeConfigParser
+import logging
+import time
+import unittest
+
+logger = logging.getLogger('example')
+logging.basicConfig(level=logging.DEBUG)
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: You don't need this. You have all configuration for logging in log.conf

File: ecomap/src/python/ecomap/config.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+class Config(object):
+
+    def __new__(cls, *args):
+        if not hasattr(cls, '_instance'):        # if there is not instance
+            cls._instance = object.__new__(cls)  # create it
+        return cls._instance                     # else return it
+
+    def __init__(self, path, logger):
+        self.config = {}                         # dictionary, contains configs
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Please add new class variable logger = logging.getLogger('example'), but with correct name:)

File: ecomap/src/python/ecomap/config.py (hunk 16 above)
Comment (padalko) [inline ecomap/src/python/ecomap/config.py]: we'll implement logging+enviroment and update it in next commit

File: ecomap/src/python/ecomap/config.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        self.update_time = time.time() + REFRESH_TIME  # time of living
+        self.path = path                         # path to file (temporary)
+        self.logger = logger
+        self.logger.debug('Initialized instance at: %s', time.time())
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: info or debug? what do you think?

File: ecomap/src/python/ecomap/config.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+                        value = eval(value)
+                    except NameError:
+                        pass
+                self.config[section + '.' + key] = value
+
+
+if __name__ == '__main__':
+    class Test(unittest.TestCase):
+
+        def test_sameinstances(self):
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: good

File: ecomap/src/python/ecomap/config.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+            self.a = Config('../../../etc/ecomap.conf', logger)
+            self.b = Config('../../../etc/ecomap.conf', logger)
+            self.assertEquals(self.a, self.b)
+
+        def test_config(self):
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: wrong

File: ecomap/src/python/ecomap/config.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+            self.a = Config('../../../etc/ecomap.conf', logger)
+            self.b = Config('../../../etc/ecomap.conf', logger)
+            self.assertEquals(self.a.get_config(), self.b.get_config())
+
+        def test_updatetime(self):
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: wrong

File: ecomap/src/python/ecomap/config.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+            self.a = Config('../../../etc/ecomap.conf', logger)
+            self.b = Config('../../../etc/ecomap.conf', logger)
+            self.assertEquals(self.a.update_time, self.b.update_time)
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: We will discuss about unittest on Tuesday.

File: ecomap/src/python/ecomap/config.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+        return self.config
+
+    def _parse_confs(self):
+        self.logger.debug('Parsed ecomap.conf at %s', (time.time()))
+        config = SafeConfigParser()              # create config object
+        config.readfp(open(self.path))           # read file
+        sections = config.sections()             # get sections
+        for section in sections:                 # for each section
+            for (key, value) in config.items(section):   # for each key/value
+                if value and key != 'password':
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Please use constant variable for 'password'.

File: ecomap/src/python/ecomap/config.py [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self, path, logger):
+        self.config = {}                         # dictionary, contains configs
+        self.update_time = time.time() + REFRESH_TIME  # time of living
+        self.path = path                         # path to file (temporary)
+        self.logger = logger
+        self.logger.debug('Initialized instance at: %s', time.time())
+
+    def get_config(self):                        # method which checks if we
+        self.logger.debug('Check if need to update at %s', time.time())
+        if self.update_time < time.time():       # need to update configs
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: wrong :)

File: ecomap/src/python/ecomap/config.py [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+            self.config = {}                     # nullify configs dictionary
+            self.update_time = time.time() + REFRESH_TIME  # set time to update
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: self.update_time = time.time()

File: ecomap/src/python/ecomap/config.py (hunk 23 above)
Comment (padalko) [inline ecomap/src/python/ecomap/config.py]: can we use here a regexp for different types of 'password' key?
for example:
import re
p = re.compile(ur'(p)(a)?s*(w)?(d?)(or)?d?$', re.IGNORECASE | re.UNICODE)
test_str = u"pass\n"
re.search(p, test_str)

File: ecomap/src/python/ecomap/config.py [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+        def test_config(self):
+            self.a = Config('../../../etc/ecomap.conf', logger)
+            self.b = Config('../../../etc/ecomap.conf', logger)
+            self.assertEquals(self.a.get_config(), self.b.get_config())
+
+        def test_updatetime(self):
+            self.a = Config('../../../etc/ecomap.conf', logger)
+            self.b = Config('../../../etc/ecomap.conf', logger)
+            self.assertEquals(self.a.update_time, self.b.update_time)
+
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Please add result from pylint :) as a comment.
Thanks.

**Comment by frutkick** [comment]
************* Module config
W: 22,12: Attribute '_instance' defined outside __init__ (attribute-defined-outside-init)
R: 27, 0: Too few public methods (1/2) (too-few-public-methods)


Report
======
40 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |1      |1          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |2      |2          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |4      |4          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|function |0      |0          |=          |0           |0        |
+---------+-------+-----------+-----------+------------+---------+



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |41     |57.75 |41       |=          |
+----------+-------+------+---------+-----------+
|docstring |23     |32.39 |23       |=          |
+----------+-------+------+---------+-----------+
|comment   |0      |0.00  |0        |=          |
+----------+-------+------+---------+-----------+
|empty     |7      |9.86  |7        |=          |
+----------+-------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |0      |0        |=          |
+-----------+-------+---------+-----------+
|refactor   |1      |1        |=          |
+-----------+-------+---------+-----------+
|warning    |1      |1        |=          |
+-----------+-------+---------+-----------+
|error      |0      |0        |=          |
+-----------+-------+---------+-----------+



Messages
--------

+-------------------------------+------------+
|message id                     |occurrences |
+===============================+============+
|too-few-public-methods         |1           |
+-------------------------------+------------+
|attribute-defined-outside-init |1           |
+-------------------------------+------------+



Global evaluation
-----------------
Your code has been rated at 9.50/10 (previous run: 9.50/10, +0.00)

Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+

File: ecomap/src/python/ecomap/config.py [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+It exists to parse *.conf files and return dictionary,
+which contains configuration from those files. Every 15 minutes
+it returns new dictionary which contains updated configs.
+"""
+import os
+import time
+
+from ConfigParser import SafeConfigParser
+
+from ecomap.utils import get_logger
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: Change on import logging

File: ecomap/src/python/ecomap/config.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Singleton class to get configs
+    """
+    __metaclass__ = Singleton
+
+    def __init__(self):
+        self.config = {}
+        self.update_time = 0
+        self.path = CONFIG_PATH
+        self.log = get_logger('config_parser')
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: change on self.log = logging.getLogger('config_parser')

File: ecomap/src/python/ecomap/config.py [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+        """
+        Call parse method if need.
+        Returns:
+            dictionary, containing configs
+        """
+        if self.update_time + REFRESH_TIME < time.time():
+            self.log.info('Refresh configs')
+            self.update_time = time.time()
+            self._parse_confs()
+        self.log.info('Return configs.')
```
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: You don't need this line:)

Comment (frutkick) [inline ecomap/src/python/ecomap/config.py]: You mean last line? :)

File: ecomap/src/python/ecomap/config.py (hunk 27 above)
Comment (padalko) [inline ecomap/src/python/ecomap/config.py]: what is right import look like?
from utils import logger 
?

File: ecomap/src/python/ecomap/config.py (hunk 29 above)
Comment (ulyktey) [inline ecomap/src/python/ecomap/config.py]: You don't need  self.log.info('Return configs.') line here. You will logging 'Return configs.' every time when you will need some variables from config file. It is very expensive.

**Pull request merged**